### DISCUSSION TITLE
Use of class TypeVariable with kwargs

### DISCUSSION THREAD

**Discussion by iyakushev** [microsoft/pyright#7301] [Q&A] [answered]
I've stumbled onto a weird behaviour when using `kwargs` in a generic class. Everything seems to work fine until I try to pass values to the initialiser as `**values`. Consider the example below. Am I mistaken with the use of `ValidParams_T` in `kwargs`?
Besides, I've noted a few oddities around positional `args`. But I presume that might be a specified corner-case which I'm not aware of.

pyright == 1.1.351

```python
class GenericValidation[ValidParams_T]:
    def __init__(self, **kwargs: ValidParams_T):
        ...

type Params = int | str | bool | list[Params] | dict[str, Params]

class SpecificValidation(GenericValidation[Params]):
    ...

values = {"key1": 1, "key2": True, "key3": [1,2,3]}

SpecificValidation(key1=1, key2=True, key3=[1,2,3])  # OK
SpecificValidation(key1=1, key2=True, key3={1,2,3})  # OK: Failed as expected

SpecificValidation(*values)  # OK: Not sure why it would let us do positional args since (below)
SpecificValidation([1])  # OK: This fails because it expects 0 positional arguemnts

SpecificValidation(**values)  # Fail: Argument of type "int | bool | list[int]" cannot be assignet to parameter "kwargs" of type "Params" in function "__init__"

```

**Comment by erictraut** [comment, marked as answer]
Pyright is working as intended for the above code.

The line where you say "Not sure why it would let us do position args" is because `values` is of type `dict[str, Unknown]`, and the length of the iteration for a `dict` is unknown. Pyright does not emit an error in this case because it could produce zero iterations. There's an argument to be made that it should assume that the length of the iterable is at least one, but that will produce false positives in some situations. When faced with a choice between false positives and false negatives, we generally err on the side of eliminating the false positive.

On the last line, there is no error in standard (default) type checking mode because the inferred type of `values` is `dict[str, Unknown]`. If you enable strict type checking mode (or more precisely, set `strictDictionaryInference` to true), then the inferred type of `values` is `dict[str, int | bool | list[int]]`. This is not compatible with the `Params` type because `list[int]` is not a subtype of `list[Params]`. If you want `values` to be assignable in this case, you'd need to declare its type as `Params` so pyright uses bidirectional type inference to override the normal type inference rules for dictionary expressions.

```python
values: Params = {"key1": 1, "key2": True, "key3": [1, 2, 3]}
```

**Reply by iyakushev** [reply]
Thank you for a swift response!
